### PR TITLE
Add EIP: Forward compatible consensus data structures

### PR DESCRIPTION
**Pull request by etan-status**

### CODE REVIEW

File: EIPS/eip-####.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
+---
```
Comment (abcoathup) [inline EIPS/eip-####.md:1]:
```suggestion
---
eip: 7688
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

File: EIPS/eip-####.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+title: Forward compatible consensus data structures
+description: Transition consensus SSZ data structures to StableContainer
+author: Etan Kissling (@etan-status)
+discussions-to: https://ethereum-magicians.org/t/eip-forward-compatible-consensus-data-structures/19673
```
Comment (abcoathup) [inline EIPS/eip-####.md:5]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-7688-forward-compatible-consensus-data-structures/19673
```

File: EIPS/eip-7688.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-04-15
+requires: 7495
+---
+
+## Abstract
+
+This EIP defines the changes needed to adopt [`StableContainer`](./eip-7495.md) in consensus data structures.
```
Comment (gaudren) [inline EIPS/eip-7688.md:16]:
```suggestion
This EIP defines the changes needed to adopt `StableContainer` from [EIP-7495](./eip-7495.md) in consensus data structures.
```

File: EIPS/eip-7688.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+While SSZ `Container` defines how data structures are merkleized, the merkleization is prone to change across the different forks. When that happens, e.g., because new features are added or old features get removed, existing verifier implementations need to be updated to be able to continue processing proofs.
+
+[`StableContainer`](./eip-7495.md) is a forward compatible alternative that guarantees a forward compatible merkleization scheme. By transitioning consensus data structures to use `StableContainer`, smart contracts that contain verifier logic no longer have to be maintained in lockstep with Ethereum's fork schedule as long as the underlying features that they verify don't change. For example, as long as the concept of slashing is represented using the boolean `slashed` field, existing verifiers will not break when unrelated features get added or removed. This is also true for off-chain verifiers, e.g., in hardware wallets or in operating systems for mobile devices that are on a different software update cadence than Ethereum.
```
Comment (gaudren) [inline EIPS/eip-7688.md:24]:
```suggestion
`StableContainer`, of [EIP-7495](./eip-7495.md), is a forward compatible alternative that guarantees a forward compatible merkleization scheme. By transitioning consensus data structures to use `StableContainer`, smart contracts that contain verifier logic no longer have to be maintained in lockstep with Ethereum's fork schedule as long as the underlying features that they verify don't change. For example, as long as the concept of slashing is represented using the boolean `slashed` field, existing verifiers will not break when unrelated features get added or removed. This is also true for off-chain verifiers, e.g., in hardware wallets or in operating systems for mobile devices that are on a different software update cadence than Ethereum.
```

**Review by gaudren — APPROVED** [review]
These aren't really that important of changes, they just fit the style better. Looks great otherwise.

File: EIPS/eip-7688.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    graffiti: Bytes32  # Arbitrary data
+    # Operations
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: Optional[SyncAggregate]
+    # Execution
+    execution_payload: Optional[ExecutionPayload]
```
Comment (wemeetagain) [inline EIPS/eip-7688.md:110]:
```suggestion
    execution_payload: Optional[StableExecutionPayload]
```

File: EIPS/eip-7688.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+- The historical forks are examined and fields that have not been present in all forks are marked as `Optional` in the new `StableContainer` type.
+- To guarantee forward and backward compatibility, new fields from future forks MUST be appended to the end of the `StableContainer` definition. Existing fields MAY be converted to `Optional`.
+
+Furthermore, a `Variant` type is defined that is specific to the fork at which the conversion is applied. This `Variant` is the equivalent of the legacy `Container` type, except that it inherits from `Variant[S]`. The SSZ serialization of `Variant` is compatible with `Container`, but the merkleization and `hash_tree_root` are computed differently. Furthermore, `Variant` MAY use fields of `Optional` type if necessary.
+
+libp2p topics still exchange fork specific `Variant` definitions using the same serialization as the legacy `Container` type. Only merkleization and `hash_tree_root` are computed differently.
```
Comment (wemeetagain) [inline EIPS/eip-7688.md:40]: Not just libp2p, but all processing will be done with Variants, right? For spec purposes, just for better type safety.

File: EIPS/eip-7688.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    previous_justified_checkpoint: Checkpoint
+    current_justified_checkpoint: Checkpoint
+    finalized_checkpoint: Checkpoint
+    # Inactivity
+    inactivity_scores: Optional[List[uint64, VALIDATOR_REGISTRY_LIMIT]]
+    # Sync
+    current_sync_committee: Optional[SyncCommittee]
+    next_sync_committee: Optional[SyncCommittee]
+    # Execution
+    latest_execution_payload_header: Optional[ExecutionPayloadHeader]
```
Comment (wemeetagain) [inline EIPS/eip-7688.md:151]:
```suggestion
    latest_execution_payload_header: Optional[StableExecutionPayloadHeader]
```

File: EIPS/eip-7688.md [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    withdrawals_root: Root
+    blob_gas_used: uint64
+    excess_blob_gas: uint64
```
Comment (wemeetagain) [inline EIPS/eip-7688.md:206]:
```suggestion
    excess_blob_gas: uint64
    ... # Additional features from the fork that introduces this EIP
```

File: EIPS/eip-7688.md [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    # Extra payload fields
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    withdrawals_root: Optional[Root]
+    blob_gas_used: Optional[uint64]
+    excess_blob_gas: Optional[uint64]
```
Comment (wemeetagain) [inline EIPS/eip-7688.md:96]:
```suggestion
    excess_blob_gas: Optional[uint64]
    ... # Additional features from the fork that introduces this EIP
```

**Review by g11tech — COMMENTED** [review]
looks great, lemme know once suggestions by @wemeetagain are incorporated, will approve the draft

File: EIPS/eip-7688.md [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Conversion procedure
+
+For each converted data structure, a new `StableContainer` type `S` is introduced that serves as the primary definition of each data structure.
+
+- Each `StableContainer` is assigned a capacity to represent its potential design space that SHALL NOT change across future forks; if it is later determined that it is insufficient, a new field can be added to contain additional fields in a sub-container.
+- The `StableContainer` starts as a copy of the latest fork's `Container` equivalent.
+- To guarantee forward and backward compatibility, new fields from future forks MUST be appended to the end of the `StableContainer` definition. Existing fields MAY be converted to `Optional`.
```
Comment (g11tech) [inline EIPS/eip-7688.md:36]:
```suggestion
- To guarantee forward and backward compatibility, new fields from future forks MUST only be appended to the `StableContainer` definition. Existing fields MAY be converted to `Optional`.
```
just a nit

**Review by g11tech — APPROVED** [review]
lgtm

**Pull request merged**